protobuf_serialization/codec.nim

```diff
@@ -0,0 +1,243 @@
+
+template toBytes*(x: fixed32 | fixed64): openArray[byte] =
+  type Base = distinctBase(typeof(x))
+  toBytesLE(Base(x))
+
+template toBytes*(x: sfixed32): openArray[byte] =
+  toBytes(fixed32(x))
+template toBytes*(x: sfixed64): openArray[byte] =
+  toBytes(fixed64(x))
+
+template toBytes*(x: pdouble): openArray[byte] =
+  cast[array[8, byte]](x)
+template toBytes*(x: pfloat): openArray[byte] =
+  cast[array[4, byte]](x)
+
+template toBytes*(header: FieldHeader): openArray[byte] =
+  toBytes(uint32(header), Leb128).toOpenArray()
+
+proc vsizeof*(x: SomeVarint): int =
+  ## Returns number of bytes required to encode integer ``x`` as varint.
+  Leb128.len(toUleb(x))
+
+proc writeValue*(output: OutputStream, value: SomeVarint) =
+  output.write(toBytes(value))
+
+proc writeValue*(output: OutputStream, value: SomeFixed64) =
+  output.write(toBytes(value))
+
+proc writeValue*(output: OutputStream, value: pstring) =
+  output.write(toBytes(puint64(string(value).len())))
+  output.write(string(value).toOpenArrayByte(0, string(value).high()))
+
+proc writeValue*(output: OutputStream, value: pbytes) =
+  output.write(toBytes(puint64(seq[byte](value).len())))
+  output.write(seq[byte](value))
+
+proc writeValue*(output: OutputStream, value: SomeFixed32) =
+  output.write(toBytes(value))
+
+proc writeField*(output: OutputStream, field: int, value: SomeScalar) =
+  output.write(toBytes(FieldHeader.init(field, wireKind(typeof(value)))))
+  output.writeValue(value)
+
+proc readValue*[T: SomeVarint](input: InputStream, _: type T): T =
+  # TODO This is not entirely correct: we should truncate value if it doesn't
+  #      fit, according to the docs:
+  #      https://developers.google.com/protocol-buffers/docs/proto#updating
+  var buf: Leb128Buf[uint64]
+  while buf.len < buf.data.len and input.readable():
+    let b = input.read()
+    buf.data[buf.len] = b
+    buf.len += 1
+    if (b and 0x80'u8) == 0:
+      break
+
+  let (val, len) = uint64.fromBytes(buf)
+  if buf.len == 0 or len != buf.len:
+    raise (ref ValueError)(msg: "Cannot read varint from stream")
+
+  fromUleb(val, T)
+
+proc readValue*[T: SomeFixed32 | SomeFixed64](input: InputStream, _: type T): T =
+  var tmp {.noinit.}: array[sizeof(T), byte]
+  if not input.readInto(tmp):
+    raise (ref ValueError)(msg: "Not enough bytes")
+  when T is pdouble | pfloat:
+    copyMem(addr result, addr tmp[0], sizeof(result))
+  elif sizeof(T) == 8:
+    cast[T](uint64.fromBytesLE(tmp)) # Cast so we don't run into signed trouble
+  else:
+    cast[T](uint32.fromBytesLE(tmp)) # Cast so we don't run into signed trouble
+
+proc readLength*(input: InputStream): int =
+  let lenu32 = input.readValue(puint32)
+  if uint64(lenu32) > uint64(int.high()):
+    raise (ref ValueError)(msg: "Invalid length")
+  int(lenu32)
+
+proc readValue*[T: SomeLengthDelim](input: InputStream, _: type T): T =
+  let len = input.readLength()
+  if len > 0:
+    type Base = T.distinctBase()
```

Comment:
```suggestion
    type Base = typetraits.distinctBase(T)
```
As this proc is generic and there's also a proc distinctBase on sugar, this can cause an issue on nim1.2